### DISCUSSION TITLE
Help migrating the secrets to the new portal method

### DISCUSSION THREAD

**Discussion by khitrenovich** [renovatebot/renovate#31707] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I'm trying to migrate the following configuration (changed URLs and package names) -

```
  "npmrc": "@designsystems:registry=https://artifactory.company.com/artifactory/api/npm/common-npm\n//artifactory.company.com/artifactory/api/npm/common-npm/:_auth=${NPM_TOKEN}\n//artifactory.company.com/artifactory/api/npm/common-npm/:always-auth=true",
  "encrypted": {
    "npmToken": "........"
  },
```

to use portal secrets according to [this sample](https://docs.renovatebot.com/getting-started/private-packages/#automatically-authenticate-for-npm-package-stored-in-private-github-npm-repository) and migration instructions.

This is what I came up with - 

```
  "hostRules": [
    {
      "matchHost": "https://artifactory.company.com/artifactory/api/npm/common-npm",
      "hostType": "npm",
      "token": "{{ secrets.COMMON_NPM_TOKEN }}"
    }
  ],
  "npmrc": "@designsystems:registry=https://artifactory.company.com/artifactory/api/npm/common-npm",
```

However, this fails with 401 error - 

```
DEBUG: GET https://artifactory.company.com/artifactory/api/npm/common-npm/@designsystems%2Fds-component-types = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=1377)
DEBUG: Failed to look up npm package @designsystems/ds-component-types
{
  "dependency": "@designsystems/ds-component-types"
  "packageFile": "client/components/package.json"
}
```

The `COMMON_NPM_TOKEN` secret was migrated in the encrypted form with the "successfully migrated" confirmation.

What did I miss here?

Thanks in advance!

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment]
See this: https://github.com/renovatebot/renovate/discussions/31253

**Reply by rarkins** [reply]
@RahulGautamSingh can you still try to fix the root cause of this?

**Reply by khitrenovich** [reply, marked as answer]
I ended up with using "raw" `npmrc` value as I had before, just replaced `${NPM_TOKEN}` with `{{ secrets.MY_NPM_TOKEN }}`.

**Reply by RahulGautamSingh** [reply]
> can you still try to fix the root cause of this?

For the below config:

```json
{
  "npmToken": "{{ secrets.NPM_TOKEN }}",
  "npmrc": "@fortawesome:registry=https://npm.fontawesome.com\n//npm.fontawesome.com/:_authToken=${NPM_TOKEN}"
}
```

Processing order within `mergeRenovateConfig()`:
a. `decryptConfig` function runs first
b. `applySecretsToConfig` function runs second

The problem:

- `decryptConfig` used to replace `${NPM_TOKEN}` in the `npmrc` string with value of `npmToken`.
- `decryptConfig` only processes items in the `encrypted` object.
- Since `npmToken` is not in the `encrypted` object anymore, `decryptConfig` doesn't touch it.
- When `applySecretsToConfig` runs, it replaces `{{ secrets.NPM_TOKEN }}` with the actual token.

The `npmToken` gets the actual token value from `secrets` object but the `${NPM_TOKEN}` substring in the `npmrc` string is never replaced with the actual token.

Still not sure why the `hostRules` approach fails though ... because the same worked for [here](https://github.com/renovatebot/renovate/discussions/31253#discussioncomment-10582777). Or is the `hostRules` approach only good if the private github npm repository and for alternative registry users need to use the `npmrc` string approach?
